src/layers/domain/ods/__init__.py

```diff
@@ -0,0 +1,79 @@
+import time
+from functools import wraps
+from http import HTTPStatus
+from typing import Callable, ParamSpec, TypeVar
+
+import requests
+from requests.exceptions import RequestException
+
+ODS_API_ENDPOINT = (
+    "https://directory.spineservices.nhs.uk/ORD/2-0-0/organisations/{ods_code}"
+)
+BACKOFF_BASE_SECONDS = 2
+
+
+class UnexpectedStatusCode(Exception):
+    pass
+
+
+class OdsApiOffline(ExceptionGroup):
+    def __new__(cls, n_attempts: int, fn_name: str, exceptions: list):
+        """
+        This is the recommended pattern for deriving from ExceptionGroup,
+        see https://docs.python.org/3/library/exceptions.html
+        """
+        self = super().__new__(
+            OdsApiOffline,
+            f"{n_attempts} exceptions were raised whilst attempting to execute '{fn_name}'",
+            exceptions,
+        )
+        return self
+
+
+def _construct_ods_url(ods_code: str) -> str:
+    return ODS_API_ENDPOINT.format(ods_code=ods_code)
+
+
+T = TypeVar("T")
+RT = TypeVar("RT")
+P = ParamSpec("P")
+
+
+def retry(max_attempts: int) -> Callable[[Callable[P, RT]], Callable[P, RT]]:
+    """Retrying on RequestException, with exponential back-off of 2*n seconds"""
+
+    def decorator(fn: Callable[P, RT]) -> Callable[P, RT]:
+        @wraps(fn)
+        def wrapper(*args: P.args, **kwargs: P.kwargs) -> Callable[P, RT]:
+            exceptions = []
+            for n in range(max_attempts):
+                try:
+                    return fn(*args, **kwargs)
+                except RequestException as exc:
+                    exceptions.append(exc)
+                finally:
+                    time.sleep(BACKOFF_BASE_SECONDS**n)
+            raise OdsApiOffline(n, fn.__name__, exceptions)
+
+        return wrapper
+
+    return decorator
+
+
+@retry(max_attempts=3)
+def is_valid_ods_code(ods_code: str) -> bool:
+    url = _construct_ods_url(ods_code=ods_code)
+    response = requests.get(url=url)
+    status_code = HTTPStatus(response.status_code)
+
+    if status_code is HTTPStatus.OK:
+        return True
+    if HTTPStatus.BAD_REQUEST <= status_code < HTTPStatus.INTERNAL_SERVER_ERROR:
+        return False
+    if status_code >= HTTPStatus.INTERNAL_SERVER_ERROR:
```

Comment:
Is there no way for the Exception group to handle these, especially if they are expected codes?

Reply:
these aren't exceptions, they're enums

Reply:
so basically its the same as:

Reply:
ok just me being dense then ha